### PR TITLE
WitnessVersion type

### PR DESCRIPTION
**Pull request by dr-orlovsky**
Reopening as a new PR to #562 (github does not allow to re-open force-pushed branches) since there might be changes during the review at this level unrelated to #563 (which will be in draft mode until this is merged - and rebased after the merge)

### CODE REVIEW

File: src/util/address.rs [hunk 1]
```diff
@@ -145,6 +154,173 @@ impl FromStr for AddressType {
...
+        })
+    }
+
+    /// Converts bitcoin script opcode into [`WitnessVersion`] variant.
+    ///
+    /// # Returns
+    /// Version of the Witness program (for opcodes in range of `OP_0`..`OP_16`)
+    ///
+    /// # Errors
+    /// If the opcode does not corresponds to any witness version, errors with
```
Comment (sgeisler) [inline src/util/address.rs:270]: s/corresponds/correspond/ ?

File: src/util/address.rs [hunk 2]
```diff
@@ -145,6 +154,173 @@ impl FromStr for AddressType {
...
+    /// [`Error::UnparsableWitnessVersion`] for the rest of opcodes
```
Comment (sgeisler) [inline src/util/address.rs:271]: "for the rest of opcodes seems redundant"

File: src/util/address.rs [hunk 3]
```diff
@@ -165,23 +341,17 @@ impl Payload {
...
         Some(if script.is_p2pkh() {
-            Payload::PubkeyHash(PubkeyHash::from_slice(&script.as_bytes()[3..23]).unwrap())
+            let mut hash_inner = [0u8; 20];
+            hash_inner.copy_from_slice(&script.as_bytes()[3..23]);
+            Payload::PubkeyHash(PubkeyHash::from_inner(hash_inner))
         } else if script.is_p2sh() {
-            Payload::ScriptHash(ScriptHash::from_slice(&script.as_bytes()[2..22]).unwrap())
+            let mut hash_inner = [0u8; 20];
+            hash_inner.copy_from_slice(&script.as_bytes()[2..22]);
+            Payload::ScriptHash(ScriptHash::from_inner(hash_inner))
```
Comment (sgeisler) [inline src/util/address.rs:350]: What is the reason for this change? Seems unrelated.

File: src/util/address.rs [hunk 4]
```diff
@@ -145,6 +154,173 @@ impl FromStr for AddressType {
...
+    }
+
+    /// Converts bitcoin script [`Instruction`] (parsed opcode) into [`WitnessVersion`] variant.
+    ///
+    /// # Returns
+    /// Version of the Witness program for [`Instruction::Op`] and [`Instruction::PushBytes`] with
+    /// byte value within `1..=16` range.
+    ///
+    /// # Errors
+    /// If the opcode does not corresponds to any witness version, errors with
```
Comment (sgeisler) [inline src/util/address.rs:288]: same as above

**Review by sgeisler — APPROVED** [review]
utACK 3645c6353757a1771799a7ec35bd1a29d34538c3

Looks pretty good. I really like the function documentation describing all error cases applicable to the function (especially when it's only a subset of the error enum).

File: src/util/address.rs [hunk 5]
```diff
@@ -56,6 +58,10 @@ pub enum Error {
     EmptyBech32Payload,
     /// Script version must be 0 to 16 inclusive
     InvalidWitnessVersion(u8),
+    /// Unable to parse witness version from string or opcode value
+    UnparsableWitnessVersion(ParseIntError),
+    /// Bitcoin script opcode does not match any known witness version, the script if malformed
```
Comment (jrawsthorne) [inline src/util/address.rs:63]:
```suggestion
    /// Bitcoin script opcode does not match any known witness version, the script is malformed
```

File: src/util/address.rs [hunk 6]
```diff
@@ -71,6 +77,8 @@ impl fmt::Display for Error {
             Error::Bech32(ref e) => write!(f, "bech32: {}", e),
             Error::EmptyBech32Payload => write!(f, "the bech32 payload was empty"),
             Error::InvalidWitnessVersion(v) => write!(f, "invalid witness script version: {}", v),
+            Error::UnparsableWitnessVersion(ref e) => write!(f, "Incorrect format of a witness version byte: {}", e),
+            Error::MalformedWitnessVersion => f.write_str("bitcoin script opcode does not match any known witness version, the script if malformed"),
```
Comment (jrawsthorne) [inline src/util/address.rs:81]:
```suggestion
            Error::MalformedWitnessVersion => f.write_str("bitcoin script opcode does not match any known witness version, the script is malformed"),
```

**Review by jrawsthorne — COMMENTED** [review]
You mentioned [here](https://github.com/rust-bitcoin/rust-bitcoin/pull/604#discussion_r634239990) about using WitnessVersion in the is_* methods on Script. Don't know if you want to do that here.

File: src/util/address.rs (hunk 3 above)
Comment (dr-orlovsky) [reply src/util/address.rs:350]: This allows to get rid of unnecessary `unwrap`

File: src/util/address.rs (hunk 1 above)
Comment (dr-orlovsky) [reply src/util/address.rs:270]:
```suggestion
    /// If the opcode does not correspond to any witness version, errors with
```

File: src/util/address.rs (hunk 2 above)
Comment (dr-orlovsky) [reply src/util/address.rs:271]:
```suggestion
    /// [`Error::UnparsableWitnessVersion`]
```

File: src/util/address.rs (hunk 4 above)
Comment (dr-orlovsky) [reply src/util/address.rs:288]:
```suggestion
    /// If the opcode does not correspond to any witness version, errors with
```

**Comment by dr-orlovsky** [comment]
@jrawsthorne well, tried that and it looks like it adds a lot more code instead of simplifying functions. So just made it for only `is_witness_program` and not the rest of `is_*` methods (which will require error testing instead of simple opcode equivalence test used now)

File: src/util/address.rs [hunk 7]
```diff
@@ -145,6 +154,173 @@ impl FromStr for AddressType {
...
+
+    /// Converts 8-bit unsigned integer value into [`WitnessVersion`] variant.
+    ///
+    /// # Returns
+    /// Version of the Witness program
+    ///
+    /// # Errors
+    /// If the integer does not corresponds to any witness version, errors with
+    /// [`Error::InvalidWitnessVersion`]
+    pub fn from_no(no: u8) -> Result<Self, Error> {
```
Comment (clarkmoody) [inline src/util/address.rs:241]: Nit: use `num` instead of `no` for a number variable? Also, this function would probably be better named `from_u8`

File: src/util/address.rs [hunk 8]
```diff
@@ -145,6 +154,173 @@ impl FromStr for AddressType {
...
+            Instruction::PushBytes(bytes) if bytes.len() == 0 => Ok(WitnessVersion::V0),
+            Instruction::PushBytes(_) => Err(Error::MalformedWitnessVersion),
+        }
+    }
+
+    /// Returns integer version number representation for a given [`WitnessVersion`] value.
+    /// NB: this is not the same as an integer representation of the opcode signifying witness
+    /// version in bitcoin script. Thus, there is no function to directly convert witness version
+    /// into a byte since the conversion requires context (bitcoin script or just a version number)
+    pub fn into_no(self) -> u8 {
```
Comment (clarkmoody) [inline src/util/address.rs:302]: Nit: I would prefer `into_u8`

**Review by clarkmoody — COMMENTED** [review]
I prefer this approach, since it "makes impossible states impossible." I've just got a couple preference nits.

File: src/util/address.rs (hunk 7 above)
Comment (dr-orlovsky) [reply src/util/address.rs:241]: It was originally named `from_u8`, but later I decided its a bad name, since there are two different u8-based representations for witness version: by script byte (`0`, `0x51`-`0x60`) and as version number (incremental from 0 to 16). So while I'm fine with using `from_num` instead of `from_no`, it will be confusing to use `from_u8`.

File: src/util/address.rs (hunk 8 above)
Comment (dr-orlovsky) [reply src/util/address.rs:302]: Commented above: the same reason for the naming here

File: src/util/address.rs [hunk 9]
```diff
@@ -145,6 +154,173 @@ impl FromStr for AddressType {
     }
 }
 
+/// Version of the WitnessProgram: first byte of `scriptPubkey` in
+/// transaction output for transactions starting with opcodes ranging from 0
+/// to 16 (inclusive).
+///
+/// Structure helps to limit possible version of the witness according to the
+/// specification; if a plain `u8` type will be used instead it will mean that
+/// version > 16, which is incorrect.
```
Comment (Kixunil) [inline src/util/address.rs:163]: Did you mean to say "if a plain `u8` type was used instead it would mean that version may be > 16, which is incorrect"?

File: src/util/address.rs [hunk 10]
```diff
@@ -145,6 +154,173 @@ impl FromStr for AddressType {
...
+    V13 = 13,
+    /// Future (unsupported) version of Witness Program
+    V14 = 14,
+    /// Future (unsupported) version of Witness Program
+    V15 = 15,
+    /// Future (unsupported) version of Witness Program
+    V16 = 16,
+}
+
+impl fmt::Display for WitnessVersion {
```
Comment (Kixunil) [inline src/util/address.rs:203]: Might be nice to add doc explaining that numeric representation is printed without any prefix (like `v`) or suffix.

File: src/util/address.rs [hunk 11]
```diff
@@ -145,6 +154,173 @@ impl FromStr for AddressType {
...
+        write!(f, "{}", *self as u8)
+    }
+}
+
+
+impl FromStr for WitnessVersion {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let version = u8::from_str(s).map_err(|err| Error::UnparsableWitnessVersion(err))?;
```
Comment (Kixunil) [inline src/util/address.rs:214]: I believe `s.parse::<u8>()` is considered idiomatic.

File: src/util/address.rs [hunk 12]
```diff
@@ -145,6 +154,173 @@ impl FromStr for AddressType {
...
+    pub fn from_instruction(instruction: Instruction) -> Result<Self, Error> {
+        match instruction {
+            Instruction::Op(op) => WitnessVersion::from_opcode(op),
+            Instruction::PushBytes(bytes) if bytes.len() == 0 => Ok(WitnessVersion::V0),
+            Instruction::PushBytes(_) => Err(Error::MalformedWitnessVersion),
+        }
+    }
+
+    /// Returns integer version number representation for a given [`WitnessVersion`] value.
+    /// NB: this is not the same as an integer representation of the opcode signifying witness
```
Comment (Kixunil) [inline src/util/address.rs:299]: I think an empty line will be needed for rustdoc to format it properly.

File: src/util/address.rs [hunk 13]
```diff
@@ -329,7 +499,7 @@ impl Address {
     }
 
     /// Check whether or not the address is following Bitcoin
-    /// standardness rules.
+    /// standard rules.
```
Comment (Kixunil) [inline src/util/address.rs:502]: Not a native speaker but the original seemed correct. The rules check if the address is standard.

**Review by Kixunil — APPROVED** [review]
Just a few doc suggestions.

File: src/util/address.rs (hunk 7 above)
Comment (Kixunil) [reply src/util/address.rs]: Possibly `raw_num` or `raw_u8`?

File: src/util/address.rs (hunk 13 above)
Comment (clarkmoody) [reply src/util/address.rs:502]: Agreed.

File: src/util/address.rs (hunk 3 above)
Comment (sgeisler) [reply src/util/address.rs:350]: Ok, but by adding a few lines and the panic would still happen if the length was wrong. I don't think it changes much, so I'm ok with it.

Comment (dr-orlovsky) [reply src/util/address.rs:350]: It would not panic since the length is checked in `is_p2sh`

File: src/util/address.rs (hunk 7 above)
Comment (sanket1729) [reply src/util/address.rs]: +1 for num instead of no

File: src/util/address.rs [hunk 14]
```diff
@@ -165,23 +341,17 @@ impl Payload {
...
-                    // Since we passed the [is_witness_program] check,
-                    // the first byte is either 0x00 or 0x50 + version.
-                    let mut verop = script.as_bytes()[0];
-                    if verop > 0x50 {
-                        verop -= 0x50;
-                    }
-                    bech32::u5::try_from_u8(verop).expect("checked before")
-                },
-                program: script.as_bytes()[2..].to_vec(),
+                version: WitnessVersion::from_opcode(opcodes::All::from(script[0])).ok()?,
```
Comment (sanket1729) [inline src/util/address.rs:353]: We should avoid opcode roundtrip with a simple `from_num`

> version: WitnessVersion::from_num(script[0]).ok()?,

File: src/blockdata/script.rs [hunk 15]
```diff
@@ -363,17 +363,17 @@ impl Script {
...
-            // Second byte push opcode 2-40 bytes
-            && self.0[1] >= opcodes::all::OP_PUSHBYTES_2.into_u8()
-            && self.0[1] <= opcodes::all::OP_PUSHBYTES_40.into_u8()
+        let script_len = self.0.len();
+        if script_len < 4 || script_len > 42 {
+            return false
+        }
+        let ver_opcode = opcodes::All::from(self.0[0]); // Version 0 or PUSHNUM_1-PUSHNUM_16
+        let push_opbyte = self.0[1]; // Second byte push opcode 2-40 bytes
+        WitnessVersion::from_opcode(ver_opcode).is_ok()
```
Comment (sanket1729) [inline src/blockdata/script.rs:372]: Can avoid opcode roundtrip by using `WitnessVersion::from_num(self.0[0]).is_ok()`

File: src/util/address.rs (hunk 14 above)
Comment (dr-orlovsky) [reply src/util/address.rs:353]: We can't since it is not a version number but an instruction byte (i.e. version 1 is encoded as `0x50`).

File: src/blockdata/script.rs (hunk 15 above)
Comment (dr-orlovsky) [reply src/blockdata/script.rs:372]: The same as above

**Comment by dr-orlovsky** [comment]
Addressed all suggestions with a separate commit

File: src/util/address.rs (hunk 14 above)
Comment (Kixunil) [reply src/util/address.rs:356]: Off-topic, just some food for thought for the future:

This code bothered me as well but couldn't point a finger on it before. The fact that this is a question strengthens it. It seems to me that `script` has more foot-gunny API than what would be ideal. I guess it'd be safer to model it similarly to `str` which doesn't implement indexing operator because it's unclear what it should return - `char` or `u8`.  Similarly I guess `script` should have explicit `.bytes()` and `.opcodes()` iterators. It'd still not prevent someone from using it incorrectly but at least correct usage would be less clunky. Not sure if there's a better way to make it safer.

File: src/util/address.rs [hunk 16]
```diff
@@ -159,8 +159,8 @@ impl FromStr for AddressType {
 /// to 16 (inclusive).
 ///
 /// Structure helps to limit possible version of the witness according to the
-/// specification; if a plain `u8` type will be used instead it will mean that
-/// version > 16, which is incorrect.
+/// specification; if a plain `u8` type was be used instead it would mean that
```
Comment (Kixunil) [inline src/util/address.rs:162]: Nit: "was ~~be~~"

File: src/util/address.rs (hunk 14 above)
Comment (sanket1729) [reply src/util/address.rs:356]: Agreed, I got hit by this footgun :)

**Review by sanket1729 — DISMISSED** [review]
ack 4ecfd689d1fa8da4aed9d9a265f73c1ef3f1160f

**Comment by dr-orlovsky** [comment]
Needed rebase, merged master into this branch so the review will not require to go through all commits again. Also fixed @Kixunil nit in docs.

Ned re-ACK from @sanket1729 and review from @apoelstra before the merge since this is intrusive and important change

File: src/util/address.rs [hunk 17]
```diff
@@ -571,15 +571,15 @@ impl fmt::Display for Address {
                     Network::Testnet | Network::Signet => "tb",
                     Network::Regtest => "bcrt",
                 };
-                let bech_ver = if ver.to_u8() > 0 {  bech32::Variant::Bech32m } else { bech32::Variant::Bech32 };
+                let bech_ver = if version.into_num() > 0 {  bech32::Variant::Bech32m } else { bech32::Variant::Bech32 };
```
Comment (Kixunil) [inline src/util/address.rs:574]: Just realized, I believe `Copy` types tend to use `to_` instead of `into_`. Not a big deal IMO.

Comment (dr-orlovsky) [reply src/util/address.rs:574]: I'm a bit confused over this, since https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv explicitly specifies that `to_*` operation is _expensive_ and takes _borrowed_ value, while `into_*` us _unexpensive_ and consumes the value, which is perfect for `Copy` types

Comment (Kixunil) [reply src/util/address.rs:574]: Looks like that docs is confusing as copying `u8` is considered inexpensive. :)

`into_` says it's for non-Copy types. :man_shrugging:

File: src/util/address.rs [hunk 18]
```diff
@@ -56,6 +58,10 @@ pub enum Error {
     EmptyBech32Payload,
     /// Script version must be 0 to 16 inclusive
     InvalidWitnessVersion(u8),
+    /// Unable to parse witness version from string or opcode value
+    UnparsableWitnessVersion(ParseIntError),
```
Comment (sgeisler) [inline src/util/address.rs:62]: Docs: is this really returned on opcode parsing failure?

**Review by sgeisler — DISMISSED** [review]
utACK 409cb919c5586b5fb39bd72dba3b73e433c379ad

File: src/util/address.rs (hunk 18 above)
Comment (dr-orlovsky) [reply src/util/address.rs:62]: Nope, only from string. Fixing

**Review by sgeisler — APPROVED** [review]
utACK 8a99a3fddce43001223774fe0d58dfc1457394bb

**Comment by apoelstra** [comment]
Can you clean up the history by removing the merge commit and rebasing on top of master?

**Comment by apoelstra** [comment]
ack 8a99a3fddce43001223774fe0d58dfc1457394bb aside from history

**Comment by dr-orlovsky** [comment]
@apoelstra done

**Review by sanket1729 — APPROVED** [review]
ACK a8571f24cb3d50289acc86a6fe72f4ecc615cee8

**Comment by dr-orlovsky** [comment]
@sanket1729 @sgeisler had to force-push due to conflicts with recent merges. Re-ACK will be highly appreciated

**Review by sgeisler — APPROVED** [review]
re-ACK ecc400826c5eebf4c8bf3f28219af348a1ba0191

Just looked at the diff-diff using `git range-diff 8a99a3f...ecc400826c5eebf4c8bf3f28219af348a1ba0191`

**Comment by apoelstra** [comment]
re-utACK ecc400826c5eebf4c8bf3f28219af348a1ba0191

**Review by apoelstra — APPROVED** [review]

**Review by sanket1729 — APPROVED** [review]
ACK ecc4008
